sbncode/PID/Dazzle_module.cc

```diff
@@ -0,0 +1,810 @@
+  float numDaughters, maxDaughterHits; // Hierarchy of the track
+
+  // MVA scores
+  float muonScore, pionScore, protonScore, otherScore, bestScore;
+  int bestPDG;
+
+  // Other metrics for filling in tree for analysis
+  TMVA::Reader* reader;
+  TTree* trackTree;
+
+  int truePdg, chi2PIDPDG, chi2PIDPDGNoKaon, numHits, bestPlane, bestPlaneHits, hierarchyDepth, trueStopping, recoContained, recoPrimary;
+
+  float chi2PIDPion, chi2PIDKaon;
+  float chi2Pol0Chi2, chi2ExpChi2;
+  // float lgcAmp, lgcGaussWidth, lgcLandauWidth, lgcChi2;
+  float stdDevDCA, maxDCA;
+  float mcsScatterMax, mcsMuonP, rangeMuonP;
+  float trackScore, daughterTrackScore;
+  float startX, startY, startZ, endX, endY, endZ, trueStartX, trueStartY, trueStartZ, trueEndX, trueEndY, trueEndZ, truePx, truePy, truePz, startDist, endDist, trueP, trueEndP, trueThetaXZ, trueThetaYZ, energyComp, energyPurity;
+
+  std::string trueType, trueEndProcess, chi2PIDType, chi2PIDTypeNoKaon;
+
+  void ClearTree();
+  void FillPFPMetrics(const art::Ptr<recob::PFParticle>& pfp, const std::map<size_t, art::Ptr<recob::PFParticle>>& pfpMap,
+      const art::FindManyP<recob::Cluster>& fmCluster, const art::FindManyP<recob::Hit>& fmHit, const art::FindManyP<larpandoraobj::PFParticleMetadata>& fmMeta);
+  void FillTrueParticleMetrics(const detinfo::DetectorClocksData& clockData, const recob::Track& track, const std::vector<art::Ptr<recob::Hit>>& hits,
+      std::vector<art::Ptr<sim::SimChannel>>& simChannels);
+  void FillTrackMetrics(const recob::Track& track);
+  void FillMCSMetrics(const recob::MCSFitResult& mcs);
+  void FillChi2PIDMetrics(const anab::ParticleID& pid);
+  void FillRangePMetrics(const RangeP& range);
+  // void FillLGCMetrics(const LGCFit& lgc);
+  void FillClosestApproachMetrics(const ScatterClosestApproach& closestApproach);
+  void FillStoppingChi2Metrics(const StoppingChi2Fit& stoppingChi2);
+  MVAPID RunMVA();
+
+  std::map<size_t, art::Ptr<recob::PFParticle>> GetPFPMap(std::vector<art::Ptr<recob::PFParticle>>& pfps) const;
+  unsigned int GetPFPHierarchyDepth(const art::Ptr<recob::PFParticle>& pfp, const std::map<size_t, art::Ptr<recob::PFParticle>>& pfpMap) const;
+  float GetPFPTrackScore(const art::Ptr<recob::PFParticle>& pfp, const art::FindManyP<larpandoraobj::PFParticleMetadata>& fmMeta) const;
+  bool InFV(const TVector3& pos) const;
+  std::string PdgString(const int pdg) const;
+};
+
+Dazzle::Dazzle(fhicl::ParameterSet const& p)
+    : EDProducer { p }
+    , fLArGeantLabel(p.get<std::string>("LArGeantLabel"))
+    , fPFPLabel(p.get<std::string>("PFPLabel"))
+    , fTrackLabel(p.get<std::string>("TrackLabel"))
+    , fCaloLabel(p.get<std::string>("CaloLabel"))
+    , fMCSLabel(p.get<std::string>("MCSLabel"), std::string("muon"))
+    , fChi2Label(p.get<std::string>("Chi2Label"))
+    , fRangeLabel(p.get<std::string>("RangeLabel"), std::string("muon"))
+    // , fLGCLabel(p.get<std::string>("LGCLabel"))
```

Comment:
I left this one in on purpose. This tool has potential but I could not get it to work both reliably an quickly enough so removed it. The producer module can be retrieved from the git history but I figured it was easier to comment out the lines here rather than have to try and recover specific lines. I don't feel particularly strongly here though and can remove it people want

Reply:
Yeah I thought this one might be more deliberate. Probably worth keeping then!